---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

# CTCF

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

Genomic coordinates of predicted CTCF binding sites with motif MA0139.1 
(Jaspar), in BED format, with strand orientation (directionality of binding). 
Human (hg19, hg38) and mouse (mm9, mm10) genomes. The binding sites were 
detected using the FIMO tool of the MEME suite using default settings. 
Extra columns include motif name (MA0139.1), score, p-value, q-value, and 
the motif sequence.

Experimental, to be submitted as an AnnotatiohHub Bioconductor package.

## Installation instructions

<!--
Get the latest stable `R` release from [CRAN](http://cran.r-project.org/). Then install `CTCF` using from [Bioconductor](http://bioconductor.org/) the following code:

```{r 'install', eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("CTCF")
```

And the development version from [GitHub](https://github.com/mdozmorov/CTCF) with:
-->

```{r 'install_dev', eval = FALSE}
BiocManager::install("mdozmorov/CTCF")
```

## Example

```{r eval = FALSE}
# hg38 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1s70dQy_zSSP8rsMa9VPEQS1DvUF-bA8o", destfile = "CTCF_hg38.RData")
load(file = "CTCF_hg38.RData")
CTCF_hg38
```
```
> CTCF_hg38
GRanges object with 56049 ranges and 5 metadata columns:
          seqnames            ranges strand |       motif     score       p.value   q.value            sequence
             <Rle>         <IRanges>  <Rle> | <character> <numeric>     <numeric> <numeric>         <character>
      [1]     chr1       11223-11241      - |    MA0139.1   24.4754 0.00000000134    0.0216 TCGCCAGCAGGGGGCGCCC
      [2]     chr1       11281-11299      - |    MA0139.1   22.7377 0.00000001010    0.0398 GCGCCAGCAGGGGGCGCTG
      [3]     chr1       24782-24800      - |    MA0139.1   17.3770 0.00000071100    0.2350 CGTCCAGCAGATGGCGGAT
      [4]     chr1       91420-91438      + |    MA0139.1   16.2951 0.00000141000    0.3080 GTGGCACCAGGTGGCAGCA
      [5]     chr1     104985-105003      - |    MA0139.1   16.7869 0.00000104000    0.2750 CCAACAGCAGGTGGCAGCC
      ...      ...               ...    ... .         ...       ...           ...       ...                 ...
  [56045]     chrY 57044316-57044334      - |    MA0139.1   16.4590 0.00000127000    0.2990 TGGTCACCTGGGGGCACTA
  [56046]     chrY 57189659-57189677      + |    MA0139.1   15.7541 0.00000195000    0.3430 TGTCCTCTAGGGGTCAGCC
  [56047]     chrY 57203409-57203427      - |    MA0139.1   15.6393 0.00000209000    0.3510 CTGCCGCAAGGGGGCGCAT
  [56048]     chrY 57215279-57215297      + |    MA0139.1   19.5738 0.00000015300    0.1190 gcgccacgagggggcggtg
  [56049]     chrY 57215337-57215355      + |    MA0139.1   24.4754 0.00000000134    0.0216 tcgccagcagggggcgccc
```

Note that the q-value cutoff is 0.5. Looking at the q-value distribution:

```{r echo=FALSE}
knitr::include_graphics('man/figures/CTCF_hg38_qvalue.png')
```

one may decide to use a more stringent cutoff. E.g., filtering by q-value less than
0.3 filters out more than half of the predicted sites. The remaining sites may be
considered as high-confidence CTCF sites.

```{r eval=FALSE}
length(CTCF_hg38)
# [1] 56049
CTCF_hg38 <- CTCF_hg38[ CTCF_hg38$q.value < 0.3 ]
length(CTCF_hg38)
# [1] 25474
```


CTCF GRanges for other organisms

```{r eval = FALSE}
# hg19 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1f0QpyzJl-cguVHmXVq-h0LNP0-I5whfZ", destfile = "CTCF_hg19.RData")
load(file = "CTCF_hg19.RData")
# mm9 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1JhwQpD8HWt9ZHrw4qDS71_CuHwrkQrWB", destfile = "CTCF_mm9.RData")
load(file = "CTCF_mm9.RData")
# mm10 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1-XZ4lDweP1MUC1nU8QGXwzfr38pgUQGt", destfile = "CTCF_mm10.RData")
load(file = "CTCF_mm10.RData")
```


Download the full data from the [Google Drive folder](https://drive.google.com/drive/folders/19ZXr7IETfks0OdYlmuc1Hqe700Pw3jPc?usp=sharing)

See [inst/scripts/make-data.R](inst/scripts/make-data.R) how to create the CTCF 
GRanges objects.


<!--
This is a basic example which shows you how to solve a common problem:

```{r example, eval = requireNamespace('CTCF')}
library("CTCF")
## basic example code
```

What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so:

```{r cars}
summary(cars)
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub!
-->

## Citation

Below is the citation output from using `citation('CTCF')` in R. Please
run this yourself to check for any updates on how to cite __CTCF__.

```{r 'citation', eval = requireNamespace('CTCF')}
print(citation("CTCF"), bibtex = TRUE)
```

Please note that the `CTCF` was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.

## Code of Conduct

Please note that the `CTCF` project is released with a [Contributor Code of Conduct](http://bioconductor.org/about/code-of-conduct/). By contributing to this project, you agree to abide by its terms.

## Development tools

* Continuous code testing is possible thanks to [GitHub actions](https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/)  through `r BiocStyle::CRANpkg('usethis')`, `r BiocStyle::CRANpkg('remotes')`, and `r BiocStyle::CRANpkg('rcmdcheck')` customized to use [Bioconductor's docker containers](https://www.bioconductor.org/help/docker/) and `r BiocStyle::Biocpkg('BiocCheck')`.
* Code coverage assessment is possible thanks to [codecov](https://codecov.io/gh) and `r BiocStyle::CRANpkg('covr')`.
* The [documentation website](http://mdozmorov.github.io/CTCF) is automatically updated thanks to `r BiocStyle::CRANpkg('pkgdown')`.
* The code is styled automatically thanks to `r BiocStyle::CRANpkg('styler')`.
* The documentation is formatted thanks to `r BiocStyle::CRANpkg('devtools')` and `r BiocStyle::CRANpkg('roxygen2')`.

For more details, check the `dev` directory.

This package was developed using `r BiocStyle::Biocpkg('biocthis')`.

## Code of Conduct

Please note that the CTCF project is released with a [Contributor Code of Conduct](http://bioconductor.org/about/code-of-conduct/). By contributing to this project, you agree to abide by its terms.
