---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

# CTCF

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

`CTCF` defines an AnnotationHub resource representing genomic coordinates 
of FIMO-predicted CTCF binding sites with motif MA0139.1 (Jaspar).

- Human (hg19, hg38) and mouse (mm9, mm10) genomes. 
- The binding sites were detected using the FIMO tool of the MEME suite using 
default settings. 
- Extra columns include motif name (MA0139.1), score, p-value, q-value, and 
the motif sequence.

## Installation instructions

Get the latest stable `R` release from [CRAN](http://cran.r-project.org/). Then 
install `CTCF` using from [Bioconductor](http://bioconductor.org/) the following 
code:

```{r 'install', eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("CTCF")
```

And the development version from [GitHub](https://github.com/mdozmorov/CTCF) with:

```{r 'install_dev', eval = FALSE}
BiocManager::install("mdozmorov/CTCF")
```

## Example

```{r get-data}
suppressMessages(library(AnnotationHub))
ah <- AnnotationHub()
query_data <- query(ah, "CTCF")
query_data
```

The FIMO-predicted CTCF sites are named as "CTCF_<genome version abbreviation>", 
e.g., "CTCF_hg38". Use `query_data <- query(ah , "CTCF_hg38")` for a more 
targeted search.

We can check the details about the object.

```{r}
query_data["AH95566"]
```

And retrieve the object.

```{r}
CTCF_hg38 <- query_data[["AH95566"]]
CTCF_hg38
```

Note that the default q-value cutoff is 0.5. Looking at the q-value distribution:

```{r echo=FALSE}
knitr::include_graphics("man/figures/CTCF_hg38_qvalue.png")
```

one may decide to use a more stringent cutoff. E.g., filtering by q-value less than
0.3 filters out more than half of the predicted sites. The remaining sites may be
considered as high-confidence CTCF sites.

```{r}
# Check length before filtering
length(CTCF_hg38)
# Filter and check length after filtering
CTCF_hg38 <- CTCF_hg38[CTCF_hg38$q.value < 0.3]
length(CTCF_hg38)
```


## CTCF GRanges for other organisms

```{r eval = FALSE}
# hg19 CTCF coordinates
CTCF_hg19 <- query_data[["AH95565"]]
# mm9 CTCF coordinates
CTCF_mm9 <- query_data[["AH95567"]]
# mm10 CTCF coordinates
CTCF_mm10 <- query_data[["AH95568"]]
```

Download the full data from the [Google Drive folder](https://drive.google.com/drive/folders/19ZXr7IETfks0OdYlmuc1Hqe700Pw3jPc?usp=sharing)

See [inst/scripts/make-data.R](inst/scripts/make-data.R) how to create the CTCF 
GRanges objects.


## Citation

Below is the citation output from using `citation('CTCF')` in R. Please
run this yourself to check for any updates on how to cite __CTCF__.

```{r 'citation', eval = requireNamespace('CTCF')}
print(citation("CTCF"), bibtex = TRUE)
```

Please note that the `CTCF` was only made possible thanks to many other R and 
bioinformatics software authors, which are cited either in the vignettes and/or 
the paper(s) describing this package.

## Development tools

* Continuous code testing is possible thanks to [GitHub actions](https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/)  through `r BiocStyle::CRANpkg('usethis')`, `r BiocStyle::CRANpkg('remotes')`, and `r BiocStyle::CRANpkg('rcmdcheck')` customized to use [Bioconductor's docker containers](https://www.bioconductor.org/help/docker/) and `r BiocStyle::Biocpkg('BiocCheck')`.
* Code coverage assessment is possible thanks to [codecov](https://codecov.io/gh) and `r BiocStyle::CRANpkg('covr')`.
* The [documentation website](http://mdozmorov.github.io/CTCF) is automatically updated thanks to `r BiocStyle::CRANpkg('pkgdown')`.
* The code is styled automatically thanks to `r BiocStyle::CRANpkg('styler')`.
* The documentation is formatted thanks to `r BiocStyle::CRANpkg('devtools')` and `r BiocStyle::CRANpkg('roxygen2')`.

For more details, check the `dev` directory.

This package was developed using `r BiocStyle::Biocpkg('biocthis')`.

## Code of Conduct

Please note that the CTCF project is released with a [Contributor Code of Conduct](http://bioconductor.org/about/code-of-conduct/). By contributing to this project, you agree to abide by its terms.
