---
title: "Introduction to CTCF"
author: 
  - name: Mikhail Dozmorov
    affiliation:
    - Virginia Commonwealth University
    email: mikhail.dozmorov@gmail.com
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('CTCF')`"
vignette: >
  %\VignetteIndexEntry{Introduction to CTCF}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL ## Related to https://stat.ethz.ch/pipermail/bioc-devel/2020-April/016656.html
)
```


```{r vignetteSetup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library("RefManageR")

## Write bibliography information
bib <- c(
    R = citation(),
    BiocStyle = citation("BiocStyle")[1],
    knitr = citation("knitr")[1],
    RefManageR = citation("RefManageR")[1],
    rmarkdown = citation("rmarkdown")[1],
    sessioninfo = citation("sessioninfo")[1],
    # testthat = citation("testthat")[1],
    CTCF = citation("CTCF")[1]
)
```

# Basics

`r Biocpkg("CTCF")` is  an experimental data package, to be submitted 
as an AnnotatiohHub Bioconductor package.

`r Biocpkg("CTCF")` contains genomic coordinates of predicted CTCF binding sites 
with motif MA0139.1 (Jaspar), in BED format, with strand orientation 
(directionality of binding). Human (hg19, hg38) and mouse (mm9, mm10) genomes. 
The binding sites were detected using the FIMO tool of the MEME suite using 
default settings. Extra columns include motif name (MA0139.1), score, p-value, 
q-value, and the motif sequence.

## Install `CTCF`

<!--`R` is an open-source statistical environment which can be easily modified to enhance its functionality via packages. `r Biocpkg("CTCF")` is a `R` package available via the [Bioconductor](http://bioconductor.org) repository for packages. `R` can be installed on any operating system from [CRAN](https://cran.r-project.org/) after which you can install `r Biocpkg("CTCF")` by using the following commands in your `R` session:-->

```{r "install", eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("mdozmorov/CTCF")

## Check that you have a valid Bioconductor installation
# BiocManager::valid()
```

## Use CTCF binding sites

```{r eval = FALSE}
# hg38 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1s70dQy_zSSP8rsMa9VPEQS1DvUF-bA8o", destfile = "CTCF_hg38.RData")
load(file = "CTCF_hg38.RData")
CTCF_hg38
```
```
> CTCF_hg38
GRanges object with 56049 ranges and 5 metadata columns:
          seqnames            ranges strand |       motif     score       p.value   q.value            sequence
             <Rle>         <IRanges>  <Rle> | <character> <numeric>     <numeric> <numeric>         <character>
      [1]     chr1       11223-11241      - |    MA0139.1   24.4754 0.00000000134    0.0216 TCGCCAGCAGGGGGCGCCC
      [2]     chr1       11281-11299      - |    MA0139.1   22.7377 0.00000001010    0.0398 GCGCCAGCAGGGGGCGCTG
      [3]     chr1       24782-24800      - |    MA0139.1   17.3770 0.00000071100    0.2350 CGTCCAGCAGATGGCGGAT
      [4]     chr1       91420-91438      + |    MA0139.1   16.2951 0.00000141000    0.3080 GTGGCACCAGGTGGCAGCA
      [5]     chr1     104985-105003      - |    MA0139.1   16.7869 0.00000104000    0.2750 CCAACAGCAGGTGGCAGCC
      ...      ...               ...    ... .         ...       ...           ...       ...                 ...
  [56045]     chrY 57044316-57044334      - |    MA0139.1   16.4590 0.00000127000    0.2990 TGGTCACCTGGGGGCACTA
  [56046]     chrY 57189659-57189677      + |    MA0139.1   15.7541 0.00000195000    0.3430 TGTCCTCTAGGGGTCAGCC
  [56047]     chrY 57203409-57203427      - |    MA0139.1   15.6393 0.00000209000    0.3510 CTGCCGCAAGGGGGCGCAT
  [56048]     chrY 57215279-57215297      + |    MA0139.1   19.5738 0.00000015300    0.1190 gcgccacgagggggcggtg
  [56049]     chrY 57215337-57215355      + |    MA0139.1   24.4754 0.00000000134    0.0216 tcgccagcagggggcgccc
```

CTCF GRanges for other organisms

```{r eval = FALSE}
# hg19 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1f0QpyzJl-cguVHmXVq-h0LNP0-I5whfZ", destfile = "CTCF_hg19.RData")
load(file = "CTCF_hg19.RData")
# mm9 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1JhwQpD8HWt9ZHrw4qDS71_CuHwrkQrWB", destfile = "CTCF_mm9.RData")
load(file = "CTCF_mm9.RData")
# mm10 CTCF coordinates
download.file(url = "https://drive.google.com/uc?export=download&id=1-XZ4lDweP1MUC1nU8QGXwzfr38pgUQGt", destfile = "CTCF_mm10.RData")
load(file = "CTCF_mm10.RData")
```

Download the full data from the [Google Drive folder](https://drive.google.com/drive/folders/19ZXr7IETfks0OdYlmuc1Hqe700Pw3jPc?usp=sharing)

See [inst/scripts/make-data.R](inst/scripts/make-data.R) how to create the CTCF 
GRanges objects.


<!--
## Required knowledge

`r Biocpkg("CTCF")` is based on many other packages and in particular in those that have implemented the infrastructure needed for dealing with RNA-seq data (EDIT!). That is, packages like `r Biocpkg("SummarizedExperiment")` (EDIT!).

If you are asking yourself the question "Where do I start using Bioconductor?" you might be interested in [this blog post](http://lcolladotor.github.io/2014/10/16/startbioc/#.VkOKbq6rRuU).

## Asking for help

As package developers, we try to explain clearly how to use our packages and in which order to use the functions. But `R` and `Bioconductor` have a steep learning curve so it is critical to learn where to ask for help. The blog post quoted above mentions some but we would like to highlight the [Bioconductor support site](https://support.bioconductor.org/) as the main resource for getting help: remember to use the `CTCF` tag and check [the older posts](https://support.bioconductor.org/t/CTCF/). Other alternatives are available such as creating GitHub issues and tweeting. However, please note that if you want to receive help you should adhere to the [posting guidelines](http://www.bioconductor.org/help/support/posting-guide/). It is particularly critical that you provide a small reproducible example and your session information so package developers can track down the source of the error.

## Citing `CTCF`

We hope that `r Biocpkg("CTCF")` will be useful for your research. Please use the following information to cite the package and the overall approach. Thank you!

```{r "citation"}
## Citation info
citation("CTCF")
```

# Quick start to using to `CTCF`

```{r "start", message=FALSE}
library("CTCF")
```

Edit this as you see fit =)

Here is an example of you can cite your package inside the vignette:

* `r Biocpkg("CTCF")` `r Citep(bib[["CTCF"]])`



# Reproducibility

The `r Biocpkg("CTCF")` package `r Citep(bib[["CTCF"]])` was made possible thanks to:

* R `r Citep(bib[["R"]])`
* `r Biocpkg("BiocStyle")` `r Citep(bib[["BiocStyle"]])`
* `r CRANpkg("knitr")` `r Citep(bib[["knitr"]])`
* `r CRANpkg("RefManageR")` `r Citep(bib[["RefManageR"]])`
* `r CRANpkg("rmarkdown")` `r Citep(bib[["rmarkdown"]])`
* `r CRANpkg("sessioninfo")` `r Citep(bib[["sessioninfo"]])`
* `r CRANpkg("testthat")` `r Citep(bib[["testthat"]])`

This package was developed using `r BiocStyle::Biocpkg("biocthis")`.


Code for creating the vignette

```{r createVignette, eval=FALSE}
## Create the vignette
library("rmarkdown")
system.time(render("CTCF.Rmd", "BiocStyle::html_document"))

## Extract the R code
library("knitr")
knit("CTCF.Rmd", tangle = TRUE)
```
-->


Date the vignette was generated.

```{r reproduce1, echo=FALSE}
## Date the vignette was generated
Sys.time()
```

Wallclock time spent generating the vignette.

```{r reproduce2, echo=FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits = 3)
```

`R` session information.

```{r reproduce3, echo=FALSE}
## Session info
library("sessioninfo")
options(width = 120)
session_info()
```



# Bibliography

This vignette was generated using `r Biocpkg("BiocStyle")` `r Citep(bib[["BiocStyle"]])`
with `r CRANpkg("knitr")` `r Citep(bib[["knitr"]])` and `r CRANpkg("rmarkdown")` `r Citep(bib[["rmarkdown"]])` running behind the scenes.

Citations made with `r CRANpkg("RefManageR")` `r Citep(bib[["RefManageR"]])`.

```{r vignetteBiblio, results = "asis", echo = FALSE, warning = FALSE, message = FALSE}
## Print bibliography
PrintBibliography(bib, .opts = list(hyperlink = "to.doc", style = "html"))
```
